sds_data_manager/lambda_code/IAlirtCode/ialirt_catalog_api.py

```diff
@@ -7,25 +13,83 @@
 
-    Blank function for initial setup.
+    Takes in a date range and ground station, and queries the s3 bucket that holds
+    the pointing schedules for ground stations part of the I-ALiRT network.
+
+    If files exist for the requested date range, the response body returns all file
+    versions for every date, including the file name, a pre-signed url for its download,
+    and its last date modified.
 
     Parameters
     ----------
     event : dict
         The JSON formatted document with the data required for the
-        lambda function to process
+        lambda function to process, i.e., date range and ground station
     context : LambdaContext
         This object provides methods and properties that provide
         information about the invocation, function,
         and runtime environment.
+
+    Returns
+    -------
+    dict
+        A dictionary containing headers, status code, and body, designed to be returned
+        as an API response.
     """
-    logger.info(f"Event: {event}")
-    logger.info(f"Context: {context}")
+    logger.info("Received event: " + json.dumps(event, indent=2))
+
+    if not event['start_date'] < event['end_date']:
+        response_body = ("The start date of the requested time frame is not before the "
+                         "end date. Please supply a valid time range.")
+        return {
+            "statusCode": 400,
+            "headers": {"Content-Type": "application/json"},
+            "body": json.dumps(response_body)
+        }
+
+    bucket = os.getenv("S3_BUCKET")
+    region = os.getenv("REGION")
+
+    s3_client = boto3.client(
+        "s3",
+        region_name=region,
+        config=botocore.client.Config(signature_version="s3v4"),
+    )
+
+    response_body = {}
+    for day in np.arange(event['start_date'], event['end_date'], dtype='datetime64[D]'):
+        day_path = (f"pointing_schedules/{event['station']}/" +
+                    day.item().strftime('%Y%m%d') + "/")
+        files = s3_client.list_objects_v2(Bucket=bucket, Prefix=day_path)
+        if 'Contents' not in files.keys():
+            response_body = (
+                f"There are not files associated with the provided date {day}. "
+                f"Please supply a valid time range that is covered by existing files.")
+            return {
+                "statusCode": 404,
+                "headers": {"Content-Type": "application/json"},
+                "body": json.dumps(response_body)
+            }
+        file_names = []
+        file_urls = []
+        dates_modified = []
+        for file in files['Contents']:
+            pre_signed_url = s3_client.generate_presigned_url(
+                "get_object",
+                Params={"Bucket": bucket, "Key": file['Key']}, ExpiresIn=604800
+            )
+            file_names.append(file['Key'].rsplit('/', 1)[1])
+            file_urls.append(pre_signed_url)
+            dates_modified.append(file['LastModified'].strftime('%Y-%m-%d %H:%M'))
```

Comment:
Out of curiosity, do they need the modified times returned in the API?

Is there a way to repurpose our current API handlers so that a user can request a specific filename and then be redirected to the file_url? So this endpoint would theoretically just return a list of filenames that the user could then go and retrieve later. Our API could send a redirect with a `Location` header in it for the presigned url, but wouldn't require the end-users to know about the details in the background.

Reply:
I paired with Jenny, who is going to be the main user of this endpoint as she builds it into the I-ALiRT website, and got some clarity around this -

Yes, we want to return the modified times when we return the file names, as they'll be displayed together.

And yes to your second point - Jenny and I agree that a better way to approach this is to break the current logic into two steps - step one is use the catalog endpoint to get a list of file names & dates modified, and step two is use the existing download endpoint to get a presigned url for a specific file.

We'll need to supply Jenny with an example url to query the catalog endpoint, and an example url to query the download endpoint within the I-ALiRT API Manager.